kernel/kernel-api/src/main/java/io/delta/kernel/internal/TableImpl.java

```diff
@@ -164,4 +168,64 @@ public CloseableIterator<ColumnarBatch> getChangesByVersion(
 
     return DeltaLogActionUtils.readCommitFiles(engine, commitFiles, readSchema);
   }
+
+  /**
+   * Returns the raw delta actions for each version between startVersion and endVersion. Only reads
+   * the actions requested in actionSet from the JSON log files.
+   *
+   * <p>For the returned columnar batches:
+   *
+   * <ul>
+   *   <li>Each row within the same batch is guaranteed to have the same commit version
+   *   <li>The batch commit versions are monotonically increasing
+   *   <li>The top-level columns include "version", "timestamp", and the actions requested in
+   *       actionSet. "version" and "timestamp" are the first and second columns in the schema,
+   *       respectively. The remaining columns are based on the actions requested and each have the
+   *       schema found in {@code DeltaAction.schema}.
+   * </ul>
+   *
+   * @param engine {@link Engine} instance to use in Delta Kernel.
+   * @param startVersion start version (inclusive)
+   * @param endVersion end version (inclusive)
+   * @param actionSet the actions to read and return from the JSON log files
+   * @return an iterator of batches where each row in the batch has exactly one non-null action and
+   *     its commit version and timestamp
+   * @throws TableNotFoundException if the table does not exist or if it is not a delta table
+   * @throws KernelException if a commit file does not exist for any of the versions in the provided
+   *     range
+   * @throws KernelException if provided an invalid version range
+   * @throws KernelException if the version range contains a version with reader protocol that is
+   *                         unsupported by Kernel
+   */
+  public CloseableIterator<ColumnarBatch> getChanges(
+      Engine engine,
+      long startVersion,
+      long endVersion,
+      Set<DeltaLogActionUtils.DeltaAction> actionSet) {
+    // Create a new action set that always contains protocol
+    Set<DeltaLogActionUtils.DeltaAction> copySet = new HashSet<>(actionSet);
+    copySet.add(DeltaLogActionUtils.DeltaAction.PROTOCOL);
+    // If protocol is not in the original requested actions we drop the column before returning
+    boolean shouldDropProtocolColumn =
+        !actionSet.contains(DeltaLogActionUtils.DeltaAction.PROTOCOL);
+
+    return getRawChanges(engine, startVersion, endVersion, copySet)
+        .map(
+            batch -> {
+              int protocolIdx = batch.getSchema().indexOf("protocol"); // must exist
```

Comment:
is this a common pattern throughout kernel code?

i.e. have a batch, get the index of a particular column, and then for each non-null value apply a function?

would it be cleaner to abstract this to a method on top of the ColumnarBatch interface that lets us apply a function for all non-null values for a particular column name?

if this occurrence is rare, then LGTM, and no need for premature optimization

Reply:
I don't think it's that common but we probably do it elsewhere once or twice. But I don't think an abstraction would necessarily belong on the ColumnarBatch interface though, maybe if we end up supporting UDF-like expressions this could be done there